commit 11a450c378bb3af6f56babbe8e88fe682b0d3c6c (refs/changes/83/683/1)
Author: Todd Whiteman <todd.whiteman@joyent.com>
Date:   2016-10-13T14:12:14-07:00 (3 years ago)
    
    DOCKER-950 docker pull fails on registry that does not use authentication

diff --git a/lib/registry-client-v2.js b/lib/registry-client-v2.js
index 41f17ab..fb026d8 100644
--- a/lib/registry-client-v2.js
+++ b/lib/registry-client-v2.js
@@ -676,7 +676,10 @@ function ping(opts, cb) {
 
     var headers = {};
     if (opts.authInfo) {
-        headers.authorization = _authHeaderFromAuthInfo(opts.authInfo);
+        var authHeader = _authHeaderFromAuthInfo(opts.authInfo);
+        if (authHeader !== undefined) {
+            headers.authorization = authHeader;
+        }
     } else if (opts.username) {
         headers.authorization = _basicAuthHeader(opts.username,
             opts.password);
@@ -999,13 +1002,16 @@ function RegistryClientV2(opts) {
     this.password = opts.password;
     this._loggedIn = false;
     this._authInfo = null;
-    this._headers = {
-        authorization: _authHeaderFromAuthInfo({
-            token: opts.token,
-            username: opts.username,
-            password: opts.password
-        })
-    };
+    this._headers = {};
+
+    var authHeader = _authHeaderFromAuthInfo({
+        token: opts.token,
+        username: opts.username,
+        password: opts.password
+    });
+    if (authHeader !== undefined) {
+        this._headers.authorization = authHeader;
+    }
 
     // XXX relevant for v2?
     //this._cookieJar = new tough.CookieJar();
@@ -1123,8 +1129,12 @@ RegistryClientV2.prototype.login = function regLogin(opts, cb) {
             assert.ok(result);
             self._loggedIn = true;
             self._authInfo = result.authInfo;
-            self._headers.authorization =
-                _authHeaderFromAuthInfo(self._authInfo);
+            var authHeader = _authHeaderFromAuthInfo(self._authInfo);
+            if (authHeader !== undefined) {
+                self._headers.authorization = authHeader;
+            } else if (self._headers.authorization) {
+                delete self._headers.authorization;
+            }
         }
         self.log.trace({err: err, loggedIn: self._loggedIn}, 'login: done');
         cb(err);
